### ISSUE TITLE
[rpmem-shuffle] pmemkv uses only hash code as the key lookup which may break the functionality when hash code conflicts

### ISSUE THREAD

**Issue by jerrychenhf** [Intel-bigdata/OAP#1537]
It seems that the underlayer pmemkv used by rpmem uses xxHash code as the sole key lookup value. 
Will this break the functionality if there are hash code conflicts?

**Comment by xuechendi** [comment]
@jerrychenhf , we are using a local copy of xxhash.hpp, so it won't need to search from system include and since this is a header instead of a lib, it will be compiled along with rpmem-shuffle. So from my point of view, there should not be any conflict if any xxhash in system exists. 
CC @Eugene-Mark  here.

**Comment by jerrychenhf** [comment]
@xuechendi  Thank you for your response.
I don't mean xxHash source code conflicts. I mean in the program logic, using hash code replacing the key may have problems when there are hash value conflicts from different keys.

like the following code:

int get_value_size(std::string &key, uint64_t* size) {
      xxh::hash64_t key_i = xxh::xxhash<64>(key);
      if (!index_map.contains(key_i)) {
        *size = 0;
        return -1;
      } else {
        struct block_meta_list* bml;
        index_map.find(key_i, bml);
        *size = bml->total_size;
        return 0;
      }
    }

I am not sure pmemkv.h under RPMem-shuffle/native/src is used or not for RPMem underlayer  pmem managment.
It it used, using hash value replacing the string key may have potential problem.

Thanks,
Haifeng

**Comment by xuechendi** [comment]
@jerrychenhf , I see, we haven't observe hash conflict so far, but yes, it is risky indeed, we may add a payload check in future. @Eugene-Mark is working on pmemkv codes refine, Eugene, can you add some todos in codes.

**Comment by Eugene-Mark** [comment]
Yes, we do use the pmemkv.h in our under layer PMem manipulation.  It's risky and we can add a Todo firstly as Chendi suggested and improve it in the future, thanks
